### ISSUE TITLE
CephFS CSI used with Kasten has issues with EAGAIN

### ISSUE THREAD

**Issue by insanemal** [ceph/ceph-csi#1761]
# Describe the bug #

Using Kasten to do backups of workloads, some backups will get stuck in the copy-out part of the job. 
During this period the application is mounting pvc's into a special copy-out pod. These mounts fail with Error 11 EAGAIN

FailedMount | MountVolume.MountDevice failed for volume "pvc-a51877a6-b630-4ee5-bc58-4bf9b11420e1" : rpc error: code = Internal desc = an error (exit status 11) occurred while running ceph args: [fs subvolume getpath kubes_ldr_cephfs csi-vol-37028e72-35ca-11eb-bb87-ca8c85f94050 --group_name csi -m ***stripped*** -c /etc/ceph/ceph.conf -n client.admin --keyfile=***stripped***]

# Environment details #

- Image/version of Ceph CSI driver : quay.io/cephcsi/cephcsi:v3.1.1
- Helm chart version : NA
- Kernel version : Linux 5.4.0-48-generic #52 SMP Sat Sep 26 08:27:15 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Mounter used for mounting PVC (for cephfs its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) :fuse
- Kubernetes cluster version : Provider: K3os Kubernetes Version: v1.19.2+k3s1
- Ceph cluster version : ceph version 15.2.5 (2c93eff00150f0cc5f106a559557a58d3d7b6f1f) octopus (stable)

# Steps to reproduce #

Run backup. 

# Actual results #

PVCs fail to mount with Error EAGAIN

# Expected behavior #

PVCs mount correctly

# Logs #

If the issue is in PVC mounting please attach complete logs of below containers.

- csi-rbdplugin/csi-cephfsplugin and driver-registrar container logs from
  plugin pod from the node where the mount is failing.

[driver-registar.log](https://github.com/ceph/ceph-csi/files/5640039/driver-registar.log)
[cephfsplugin.log](https://github.com/ceph/ceph-csi/files/5640041/cephfsplugin.log)

This was the most recent mount attempts.

**Comment by Madhu-1** [comment]
@insanemal can you run `ceph fs subvolume getpath kubes_ldr_cephfs csi-vol-37028e72-35ca-11eb-bb87-ca8c85f94050 --group_name csi` in your ceph cluster and see the subvolume exists

**Comment by insanemal** [comment]
Yeah I did try that it returns the same error EAGAIN saying the specified subvol is not ready. 

From a different node if I mount up the cephFS filesystem I can browse to the volume subfolder and locate this particular one, however it's empty. 

So I can only assume something has gone wrong at some point

**Comment by insanemal** [comment]
But current theory is that the previous out of space event, that was triggered by the bug from my previous ticket has left the filesystem in a inconsistent state.

**Comment by Madhu-1** [comment]
I don't see an issue from the cephcsi side, if the subvolume is not ready in the backend cephcsi cannot do anything.

**Comment by insanemal** [comment]
So the issue is, the previous version of cephCSI had a bug with snapshots. That has been fixed I believe.
I hit that bug at the same time as running out of space.

So now I have a bunch of stuck subvolumes. They can't be deleted because they have snapshots. The snapshots can't be deleted because they have pending clones. The clones aren't happening because of the previous bugs/ran out of space issue.

And I can't figure out how to cancel/delete the 'in progress' clones. 

Basically you are both correct and not quite correct. I ended up in this situation because the CephCSI can't handle things failing and/or doesn't clean up after failures correctly.

So yeah, my filesystem appears to be kinda trashed and yeah CephCSI did it. And will probably do it again because it doesn't handle failures well

**Comment by Madhu-1** [comment]
> So the issue is, the previous version of cephCSI had a bug with snapshots. That has been fixed I believe.
> I hit that bug at the same time as running out of space.
> 

Yes the bug is fixed now

> So now I have a bunch of stuck subvolumes. They can't be deleted because they have snapshots. The snapshots can't be deleted because they have pending clones. The clones aren't happening because of the previous bugs/ran out of space issue.
> 
> And I can't figure out how to cancel/delete the 'in progress' clones.
> 

You have an option in ceph to cancel ongoing clone see https://docs.ceph.com/en/latest/cephfs/fs-volumes/#cloning-snapshots  

> Basically you are both correct and not quite correct. I ended up in this situation because the CephCSI can't handle things failing and/or doesn't clean up after failures correctly.
> 
> So yeah, my filesystem appears to be kinda trashed and yeah CephCSI did it. And will probably do it again because it doesn't handle failures well

 Best option is now to clean up the stale snapshots and clones.

**Comment by insanemal** [comment]
To cancel an ongoing clone you need the clone name. And how does one find the clone name of a clone that didn't start/finish?
It's not stored in the .meta file. 
I can't find it listed in the subvolume listing "ceph fs subvolume ls blah blah"

So yeah that doesn't really work. It looks like the volume might have been cleaned up due to a delete retention policy. 

An while that bug I hit might be fixed, I'm still not convinced that things will be handled correctly if you run out of space during a snapshot clone. 

But please, if you can find actually helpful documentation on how to unravel this Gordian knot. I'm sorry if I sound frustrated, I've got considerable ceph experience. I know I should "just" clean up the stale state, but that is not working.

I haven't even mentioned the stuck k8s objects that won't delete and don't exist on the back end. No force deleting them also fails. 

I'm a tad frustrated as there are some parts of this that feel distinctly half sucked. Like not being able to remount a RWX volume if it was mounted and a k8s node crashes without cleaning unmounting it. And that being deemed "compliant with K8s spec"

So I do apologise if I seem short, but I've just had this plug in potentially cause issues in my filesystem in a way that currently looks irreparable and I'm not convinced there is sufficient edge case testing, in part due to the overwhelming feeling from comments that seem to suggest that some k8s developers (and plugin developers) think that node failure and space exhaustion are things that only happen in imaginary lands that aren't production.

**Comment by insanemal** [comment]
```
root@mgr:/mnt/ceph/volumes/csi$ceph fs subvolume snapshot rm kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi                                                                                                     
Error EAGAIN: snapshot 'csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb' has pending clones
root@mgr:/mnt/ceph/volumes/csi$ ls -lah ../_index/clone/b94869cc-9147-4bc3-aa8b-00d21237d94c
lrwxrwxrwx 1 root root 57 Dec  4 10:00 ../_index/clone/b94869cc-9147-4bc3-aa8b-00d21237d94c -> /volumes/csi/csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050
root@mgr:/mnt/ceph/volumes/csi$ ls csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050
ls: cannot access csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050: No such file or directory
root@mgr:/mnt/ceph/volumes/csi$ 
```

Yeah exactly what I thought was the case. The 'target volume' doesn't exist anymore (or never did) so I can't actually rm the snapshots because "pending clones"

Yeah this is probably an upstream CephFS issue, but its something you can encounter with the CSI plugin. 

Also it's causing me no end of issue because of the huge count of in progress clones it does know about taking up all the simultaneous clone slots. 

And look I know I sound cranky but there seems to be a lot of "well its their fault" going around.

**Comment by Madhu-1** [comment]
Okay we need to remove the stale subvolume created in the backend, we can get the list of subvolumes created by csi diver with

`kubectl get pv -o=jsonpath='{.items[*].spec.csi.volumeAttributes.subvolumeName}'`

make sure all PV has the subvolumeName in the spec. Once we get the list we need to get all created subvolumes list from  the ceph cluster
`ceph fs subvolume ls myfs csi  -> replace myfs with the filesystem name.`

Just do a comparison and if you see any subvolume present in the ceph cluster, not in the kubernetes cluster we can remove it with force command.

If its a clone check the clone is in completed status, if not and it's a stale clone we need to first cancel the clone and then remove it

Hope we have cleaned up the clones/stale subvolume we need to remove the snapshot on the subvolume. can you try to remove the snapshots with --force flag which might help to clean up snapshot if the clone is not present.

**Comment by insanemal** [comment]
You would hope so, but alas that is not the case
```
 ceph fs subvolume snapshot rm kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi --force
Error EAGAIN: snapshot 'csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb' has pending clones
```

I wasn't kidding when I said irreparable. I wouldn't have opened a ticket if I could have solved it with a few sternly worded rm commands.

**Comment by insanemal** [comment]
This is also why I'm going to be pretty insistent on some corner case detection for this particular bunch of fun. 

This apparently isn't super hard to get yourself into a state where even manual intervention isn't possible and there is nothing like it automatically making thing far worse during what is already a major outage event

**Comment by Madhu-1** [comment]
@ShyamsundarR @kotreshhr can you please help here?

**Comment by ShyamsundarR** [comment]
@insanemal As the Ceph cluster is lacking space (as per your troubleshooting) to complete operations, and due to the earlier CSI snapshot and clone leak, cleaning this up becomes more involved as clones would never complete.

We/I do acknowledge the fact that in progress clones have no way of being tracked from kubernetes to CephFS as things stand, but more on this later. Let's attempt to address cleaning up the pending operations and resources first.

The following (lengthy) procedure should help with cleaning up the orphaned and stuck resources,

1) Update to the latest ceph-csi images, if not already done, so that we can prevent further leaks

2) As clone is occurring asynchronously, just deleting the clone in progress subvolumes from CephFS will only trigger more clone creations from the kubernetes/ceph-csi front. To stop this from happening,
  - Scale down the ceph-csi provisioner Deployment to 0 (preventing any further dynamic provisioning operations)

3) As clones in progress do not have a usable CSI volumeHandle to find the subvolume that is the clone target, we would need to do the following,
  - Determine all bound PVCs from the kubernetes cluster
  - Determine all the subvolumes in CephFS
  - All subvolumes that are not traceable back to a bound PVC are possible clones or leaked clones (orphan subvolumes), this is the set we would work with going forward
  - Ensure that the orphan subvolumes are clones, inspecting their info/status outputs
    - If they are not, these may need more attention as we do not leak volumes
  - Cancel the clone operation for these subvolumes, if clone is still pending/inprogress
  - Remove these orphaned subvolumes

4) We would need to cleanup orphan snapshots in a similar manner
  - Determine all VolumeSnapshots from the kubernetes cluster
  - Determine all snapshots of subvolumes in CephFS
  - Again determine the orphan snapshot list
  - Unprotect the orphan snapshot
    - Explicit snapshots protection was removed as a per-requisite in Octopus release that you are using, but an unprotect does not hurt. As the pending clones are cancelled, snapshot unprotect or delete should work at this point.
  - Delete the orphan snapshot

5) Cleanup kubernetes state
  - Delete all PVCs that are not bound
    - This maybe optional, but if kasten has unbound PVCs due to backup requests generated earlier, it would be better to delete them to avoid recreating the same
  - Delete all VolumeSnapshots that are not ready
    - NOTE: Kubernetes VolumeSnapshot finalizers may prevent this deletion, in which case these may need deletion post step (6) once they report that they are ready or have a backing VolumeSnapshotContent

At this point the Ceph cluster should have recovered space and also not have any orphan resources, neither should the kubernetes cluster attempt creation of older PVCs/Snapshots as these are also deleted.

6) Scale up the provisioner Deployment (back to 3) for future provisioning requests

**Comment by insanemal** [comment]
@ShyamsundarR 
Sorry I wasn't super clear. It's no longer out of space. I've fixed that quite some time ago.
The issues I'm facing are from when it ran out of space while colliding at high speed with the bug that has been fixed. 

I'm on the latest stable cephfs and ceph rbd CSI plugins.  That was first cab off the rank. 

I don't have any more clones happening. I've already stopped that from happening. 

The issue is, I have a volume with snapshots I cannot delete because they apparently have pending clones. And I cannot cancel the pending clones because they don't exist. 

Unprotect does not work. 

```
root@mgr:/mnt/ceph/volumes/csi$ ceph fs subvolume snapshot unprotect kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi                                                                                              
Deprecation warning: 'snapshot unprotect' call is deprecated and will be removed in a future release
root@mgr:/mnt/ceph/volumes/csi$ ceph fs subvolume snapshot rm kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi
Error EAGAIN: snapshot 'csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb' has pending clones
```

I know how to locate what is and isn't in K8s. I know how to do all of this stuff. I'm not trying to be rude here but nobody seems to understand that. 

This has damaged my cephfs volume in a way that apparently cannot be undone. I would agree that the bug you have fixed might have resolved some of the issue. 

And I'm starting to feel that this is an upstream issue with ceph/cephfs. But far as I can tell I'm hosed. I cannot untangle this mess. These operations should be atomic. This should not be possible. 

Thanks for your assistance despite my frustration.

**Comment by ShyamsundarR** [comment]
> The issue is, I have a volume with snapshots I cannot delete because they apparently have pending clones. And I cannot cancel the pending clones because they don't exist.
> 
> Unprotect does not work.
> 
> ```
> root@mgr:/mnt/ceph/volumes/csi$ ceph fs subvolume snapshot unprotect kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi                                                                                              
> Deprecation warning: 'snapshot unprotect' call is deprecated and will be removed in a future release
> root@mgr:/mnt/ceph/volumes/csi$ ceph fs subvolume snapshot rm kubes_ldr_cephfs csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0 csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb csi
> Error EAGAIN: snapshot 'csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb' has pending clones
> ```

There can be 2 cases here,
1. There really is a pending clone that is stuck
2. There is stale state for a clone that disallows the snapshot from being deleted

To determine what is the actual state if you could do the following, it may help taking this up further,
  - For the above snapshot that you presented, the UUID of the clone that is pending can be determined using the `.meta` file within the subvolume which should be (approximately) at, `/volumes/csi/csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0`. If you `cat` this file it should have sections one of which would be `clone snaps` which contains the required UUID to investigate in `/volumes/clones/<UUID>`
  - A list of pending clones can be seen as symlinks within the CephFS file system at `/volumes/clones/<symlinks>` where each of the symlinks are named using a UUID and points to a subvolume that is being cloned

Chasing down the above for the example you presented, may help in understanding why and where a clone is stuck (if at all) and further clone state to determine what else is incorrect and needs attention.

**Comment by insanemal** [comment]
@ShyamsundarR 
Yep I know and I already did that. Sorry you either missed it or I didn't explain it well enough. I checked the meta file then checked the clones folder... and

```
root@mgr:/mnt/ceph/volumes/csi/csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0$ cat .meta 
[GLOBAL]
version = 1
type = subvolume
path = /volumes/csi/csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0/4c14b7d2-abbe-4a67-ae26-aaa1c1111521
state = complete

[clone snaps]
b94869cc-9147-4bc3-aa8b-00d21237d94c = csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb
ccb1ad88-d049-482d-a093-6d2bdf45a7d7 = csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb
bc15c6e8-3b88-4342-b014-8e5f88d2dfdd = csi-snap-b9d57cb3-35c3-11eb-bea0-7e65af608dfb
130026d8-1cbd-47e2-b999-342d452a1135 = csi-snap-e45d5f9a-368c-11eb-bea0-7e65af608dfb
a91a889c-5d4b-42da-b33a-826406eb565a = csi-snap-e45d5f9a-368c-11eb-bea0-7e65af608dfb
c6d0f4a1-936d-4be6-b22d-44c2f090464f = csi-snap-e45d5f9a-368c-11eb-bea0-7e65af608dfb
0865d464-4fbc-46a9-8d09-dae4b8611ac5 = csi-snap-d1afa2b6-3755-11eb-bea0-7e65af608dfb
24d7cd77-1885-4181-a797-4ac9c4fc238e = csi-snap-d1afa2b6-3755-11eb-bea0-7e65af608dfb
4a8c62c4-213e-4dd7-81b2-1da74916562e = csi-snap-d1afa2b6-3755-11eb-bea0-7e65af608dfb
a80ae340-6bea-4534-9799-dc9bc145d7eb = csi-snap-d6b5f01b-381e-11eb-bea0-7e65af608dfb
ab116728-3a4a-438a-b8f5-7fa1f986b170 = csi-snap-d6b5f01b-381e-11eb-bea0-7e65af608dfb

root@mgr:/mnt/ceph/volumes/csi/csi-vol-025051ea-19e3-11eb-9f05-ca15dfd3b3f0$ cd ..

root@mgr:/mnt/ceph/volumes/csi$ ls -lah ../_index/clone/b94869cc-9147-4bc3-aa8b-00d21237d94c
lrwxrwxrwx 1 root root 57 Dec  4 10:00 ../_index/clone/b94869cc-9147-4bc3-aa8b-00d21237d94c -> /volumes/csi/csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050

root@mgr:/mnt/ceph/volumes/csi$ ls csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050
ls: cannot access csi-vol-c8fbf801-35c3-11eb-bb87-ca8c85f94050: No such file or directory
```

They don't exist! I mean they have left marks all over the file system but ultimately something is very VERY wrong here

**Comment by ShyamsundarR** [comment]
> @ShyamsundarR
> Yep I know and I already did that. Sorry you either missed it or I didn't explain it well enough. I checked the meta file then checked the clones folder... and

> They don't exist! I mean they have left marks all over the file system but ultimately something is very VERY wrong here

Thank you, I did miss the fact that you checked the clones symlinks and the meta file as in this [comment](https://github.com/ceph/ceph-csi/issues/1761#issuecomment-740464531).

Can you let me know if other clone symlinks other than `b94869cc-9147-4bc3-aa8b-00d21237d94c` also do not exist? I would like to ensure the problem is stale symlinks and hence snapshot can never be deleted.

Did you delete the clone pending subvolumes using `rmdir <mountpoint>/volumes/csi/csi-vol-<uuid>` or use the subvolume series of commands to delete the subvolume? If the former, then the subvolume mgr module would not have had a chance to perform the index cleanup and could explain the issue.

Also, is the problem for other subvolumes as well, or only this subvolume?

I am not sure when all the events transpired, but if we could get the ceph-mgr logs it may help unravel the bug better.

If all boxes check off and we are looking at a bug in CephFS subvolume clone module, for recovery I would suggest removing the index entries and also cleaning up the `clone snaps` section of the `.meta` file. This would enable deleting the snapshots, as no further meta entries exist and also the index is empty/clean.

**Comment by insanemal** [comment]
@ShyamsundarR I don't believe any of the subvols were "rm -rf"ed

This does however give me a way to move forward. 

This isn't the only subvolume it's just the one I've had the most "Can't delete snapshot" errors in kubernetes about. 

There are at least 3-4 others. 

I can try and get you the mgr logs if it helps.

I want to be clear, I'm more worried about this being an unusual corner case when you run out of space while attempting to take snapshots. So not a straight out bug more an unhandled edge case. 

But I try and get logs for your. And might have a crack at undoing the issue

**Comment by insanemal** [comment]
@ShyamsundarR I appear to have untangled the issue. But have a new one. I have snapshotcontent objects that don't have snapshots on the filesystem.  I cannot delete these kubernetes objects because they fail to delete the ceph object with 

`E1218 04:00:52.375920       1 goroutinemap.go:150] Operation for "delete-snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11" failed. No retries permitted until 2020-12-18 04:02:54.375893038 +0000 UTC m=+772050.178120189 (durationBeforeRetry 2m2s). Error: "failed to delete snapshot \"snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11\", err: failed to delete snapshot content snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11: \"rpc error: code = Internal desc = volume not found\""`

--force doesn't work either.

**Comment by ShyamsundarR** [comment]
> @ShyamsundarR I appear to have untangled the issue. But have a new one. I have snapshotcontent objects that don't have snapshots on the filesystem. I cannot delete these kubernetes objects because they fail to delete the ceph object with
> 
> `E1218 04:00:52.375920 1 goroutinemap.go:150] Operation for "delete-snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11" failed. No retries permitted until 2020-12-18 04:02:54.375893038 +0000 UTC m=+772050.178120189 (durationBeforeRetry 2m2s). Error: "failed to delete snapshot \"snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11\", err: failed to delete snapshot content snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11: \"rpc error: code = Internal desc = volume not found\""`
> 
> --force doesn't work either.

@insanemal So the CephFS snapshots do not exist, or maybe the subvolume itself does not exist on the CephFS volume, would that be a correct assumption?

@Madhu-1 Delete should succeed if the storage objects are no longer present, as per the CSI protocol, right? So if the above is true, can you check if we return success for a DeleteSnapshot in such cases, thanks!

**Comment by insanemal** [comment]
@ShyamsundarR @Madhu-1 
Snapshots do not exist. For at least one of them the subvolume also does not exist. 

I haven't had a chance to see if the subvolume is gone for the few left stuck in k8s. I've been chasing other wupuses

**Comment by Madhu-1** [comment]
> > @ShyamsundarR I appear to have untangled the issue. But have a new one. I have snapshotcontent objects that don't have snapshots on the filesystem. I cannot delete these kubernetes objects because they fail to delete the ceph object with
> > `E1218 04:00:52.375920 1 goroutinemap.go:150] Operation for "delete-snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11" failed. No retries permitted until 2020-12-18 04:02:54.375893038 +0000 UTC m=+772050.178120189 (durationBeforeRetry 2m2s). Error: "failed to delete snapshot \"snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11\", err: failed to delete snapshot content snapcontent-5f6e1cb0-c123-4ce7-9fa2-2b1052787f11: \"rpc error: code = Internal desc = volume not found\""`
> > --force doesn't work either.
> 
> @insanemal So the CephFS snapshots do not exist, or maybe the subvolume itself does not exist on the CephFS volume, would that be a correct assumption?
> 
> @Madhu-1 Delete should succeed if the storage objects are no longer present, as per the CSI protocol, right? So if the above is true, can you check if we return success for a DeleteSnapshot in such cases, thanks!

Looks like it's the same as https://github.com/ceph/ceph-csi/issues/1647
